File: src-ui-react/src/lib/ipc.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

// ── Event channel names ──────────────────────────────────────────────────────
export const CHAT_EVENT = "chat://event";
export const AGENT_EVENT = "agent://event";
export const CONFIG_CHANGED_EVENT = "config://changed";
export const SERVER_LOG_EVENT = "server://log";

// ── Shared types (mirrors crates/shared/src/ipc.rs) ─────────────────────────
export interface ChatMessage {
  role: string;
  content: string;
}

export interface ChatRequest {
  stream_id: string;
  host: string;
  port: number;
  model: string;
  messages: ChatMessage[];
  temperature: number;
}

export interface ModelList {
  online: boolean;
  models: string[];
}

export type ChatEvent =
  | { kind: "token"; stream_id: string; delta: string }
  | { kind: "done"; stream_id: string }
  | { kind: "error"; stream_id: string; message: string };

export interface PlanStep {
  id: string;
  description: string;
  status: "pending" | "running" | "done" | "failed";
}

export interface ToolCall {
  call_id: string;
  tool: string;
  args: unknown;
  sensitive: boolean;
}

export interface ToolResult {
  call_id: string;
  ok: boolean;
  output: string;
}

export interface ApprovalRequest {
  call_id: string;
  tool: string;
  args: unknown;
  reason: string;
}

export interface AgentRequest {
  host: string;
  port: number;
  model: string;
  task: string;
}

export interface ApprovalDecision {
  agent_id: string;
  call_id: string;
  approved: boolean;
}

export type AgentEvent =
  | { kind: "started"; agent_id: string; parent: string | null; role: string; task: string }
  | { kind: "plan"; agent_id: string; steps: PlanStep[] }
  | { kind: "plan_update"; agent_id: string; step_id: string; status: string }
  | { kind: "step"; agent_id: string; index: number; thought: string }
  | { kind: "token"; agent_id: string; delta: string }
  | { kind: "tool_call"; agent_id: string; call: ToolCall }
  | { kind: "tool_result"; agent_id: string; result: ToolResult }
  | { kind: "approval_request"; agent_id: string; request: ApprovalRequest }
  | { kind: "sub_agent_spawned"; agent_id: string; child_id: string; role: string; task: string }
  | { kind: "log"; agent_id: string; message: string }
  | { kind: "done"; agent_id: string; final_text: string }
  | { kind: "error"; agent_id: string; message: string };

export interface TodoItem {
  id: string;
  text: string;
  done: boolean;
  priority: string;
  due_date: string | null;
  tags: string[];
}

export interface Note {
  name: string;
  content: string;
  tags: string[];
  pinned: boolean;
}

export interface NotesStore {
  notes: Note[];
}

export interface CalendarEvent {
  id: string;
  title: string;
  time: string;
  prompt: string;
  note: string | null;
  color: string | null;
  status: "pending" | "firing" | "completed" | "failed";
  result: string | null;
  tags: string[];
  recurrence: string | null;
}

export interface CalendarState {
  events: CalendarEvent[];
}

export interface KanbanTask {
  id: string;
  title: string;
  summary: string;
  assigned_agent: string;
  status: "backlog" | "todo" | "in_progress" | "done";
  priority: string;
  created_at: string;
}

export interface PlannerState {
  tasks: KanbanTask[];
}

// ── ServerConfig (partial — all UI-relevant fields) ──────────────────────────
export interface ServerConfig {
  exe_path: string;
  model_path: string;
  model_dir: string;
  model_alias: string;
  model_url: string;
  hf_repo: string;
  hf_file: string;
  chat_template: string;
  system_prompt: string;
  host: string;
  port: number;
  ctx_size: number;
  predict: number;
  gpu_layers: number;
  threads: number;
  threads_batch: number;
  threads_http: number;
  batch_size: number;
  ubatch_size: number;
  temperature: number;
  top_k: number;
  top_p: number;
  flash_attn: boolean;
  cont_batching: boolean;
  parallel: number;
  searxng_url: string;
  // UI theme
  dark_mode: boolean;
  theme_name: string;
  ui_transparency: number;
  ui_background_color: string;
  ui_text_color: string;
  ui_accent_color: string;
  ui_card_bg: string;
  ui_sidebar_bg: string;
  ui_border_color: string;
  ui_button_bg: string;
  ui_button_text: string;
  ui_card_text: string;
  ui_font_family: string;
  ui_blur: boolean;
  ui_blur_intensity: number;
  ui_radius_sm: string;
  ui_radius_md: string;
  ui_radius_lg: string;
  ui_radius_xl: string;
  ui_light_bg: string;
  ui_light_text: string;
  ui_light_accent: string;
  ui_light_card_bg: string;
  ui_light_sidebar_bg: string;
  ui_light_border: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

// ── Command wrappers ─────────────────────────────────────────────────────────
export const ipc = {
  getConfig: () => invoke<ServerConfig>("get_config"),
  updateConfig: (config: ServerConfig) => invoke<void>("update_config", { config }),

  chatListModels: (host: string, port: number) =>
    invoke<ModelList>("chat_list_models", { host, port }),
  chatSend: (req: ChatRequest) => invoke<void>("chat_send", { req }),
  chatSaveHistory: (history: ChatMessage[]) =>
    invoke<void>("chat_save_history", { history }),
  chatLoadHistory: () => invoke<ChatMessage[]>("chat_load_history"),

  agentStart: (req: AgentRequest) => invoke<string>("agent_start", { req }),
  agentApprove: (decision: ApprovalDecision) =>
    invoke<void>("agent_approve", { decision }),
  agentCancel: (agentId: string) => invoke<void>("agent_cancel", { agentId: agent_id }),
  agentStatus: () => invoke<string[]>("agent_status"),

  serverStart: () => invoke<void>("server_start"),
  serverStop: () => invoke<void>("server_stop"),
  serverStatus: () => invoke<boolean>("server_status"),
  pickPath: (directory: boolean) => invoke<string | null>("pick_path", { directory }),

  todosGet: () => invoke<TodoItem[]>("todos_get"),
  todosSet: (items: TodoItem[]) => invoke<void>("todos_set", { items }),

  notesGet: () => invoke<string>("notes_get"),
  notesSet: (content: string) => invoke<void>("notes_set", { content }),
  notesStoreGet: (scope: string) => invoke<NotesStore>("notes_store_get", { scope }),
  notesStoreSet: (scope: string, store: NotesStore) =>
    invoke<void>("notes_store_set", { scope, store }),

  calendarLoad: () => invoke<CalendarState>("calendar_load"),
  calendarSave: (events: CalendarEvent[]) =>
    invoke<void>("calendar_save", { events }),
  calendarAddEvent: (event: CalendarEvent) =>
    invoke<CalendarState>("calendar_add_event", { event }),
  calendarDeleteEvent: (id: string) =>
    invoke<CalendarState>("calendar_delete_event", { id }),

  plannerLoad: () => invoke<PlannerState>("planner_load"),
  plannerSave: (tasks: KanbanTask[]) => invoke<void>("planner_save", { tasks }),

  winMinimize: () => invoke<void>("win_minimize"),
  winToggleMaximize: () => invoke<void>("win_toggle_maximize"),
  winClose: () => invoke<void>("win_close"),
};

// Fix the agentCancel typo above
ipc.agentCancel = (agentId: string) => invoke<void>("agent_cancel", { agent_id: agentId });

// ── Typed event listeners ────────────────────────────────────────────────────
export const onChatEvent = (cb: (ev: ChatEvent) => void): Promise<UnlistenFn> =>
  listen<ChatEvent>(CHAT_EVENT, (e) => cb(e.payload));

export const onAgentEvent = (cb: (ev: AgentEvent) => void): Promise<UnlistenFn> =>
  listen<AgentEvent>(AGENT_EVENT, (e) => cb(e.payload));

export const onConfigChanged = (cb: (cfg: ServerConfig) => void): Promise<UnlistenFn> =>
  listen<ServerConfig>(CONFIG_CHANGED_EVENT, (e) => cb(e.payload));

export const onServerLog = (cb: (line: string) => void): Promise<UnlistenFn> =>
  listen<string>(SERVER_LOG_EVENT, (e) => cb(e.payload));

File: crates/shared/src/ipc.rs
//! Data-transfer objects exchanged over the Tauri IPC bridge.
//!
//! Two kinds of traffic (see [[migration-conventions]]):
//!   * request/response — command argument + return types (`*Request` / `ModelList` / …)
//!   * backend → frontend streaming — the [`ChatEvent`] and [`AgentEvent`] enums.
//!
//! We stream a single *tagged* enum per domain on one event channel
//! (`chat://event`, `agent://event`) rather than many event names. The frontend
//! registers one listener and `match`es on the tag — fewer moving parts, and new
//! event variants don't require new listeners.

use serde::{Deserialize, Serialize};

/// Event channel names (the single source of truth for both sides).
pub const CHAT_EVENT: &str = "chat://event";
pub const AGENT_EVENT: &str = "agent://event";
pub const CONFIG_CHANGED_EVENT: &str = "config://changed";
/// Each llama-server stdout/stderr line is emitted as a plain `String` here.
pub const SERVER_LOG_EVENT: &str = "server://log";

// ── Chat ──────────────────────────────────────────────────────────────────

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ChatMessage {
    pub role: String, // "system" | "user" | "assistant"
    pub content: String,
}

// ── Productivity (todos / notes) ─────────────────────────────────────────────

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct TodoItem {
    pub id: String,
    pub text: String,
    pub done: bool,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Default)]
pub struct Note {
    pub name: String,
    pub content: String,
}

#[derive(Clone, Debug, Serialize, Deserialize, Default, PartialEq)]
pub struct NotesStore {
    pub notes: Vec<Note>,
}

/// Arguments for the `chat_send` command. `host`/`port` target either the local
/// server or a routed instance; `stream_id` correlates the streamed events back
/// to the in-flight assistant message in the UI.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ChatRequest {
    pub stream_id: String,
    pub host: String,
    pub port: u16,
    pub model: String,
    pub messages: Vec<ChatMessage>,
    pub temperature: f32,
}

/// Return type of `chat_list_models`.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct ModelList {
    pub online: bool,
    pub models: Vec<String>,
}

/// Streamed on [`CHAT_EVENT`]. `stream_id` lets the UI route deltas to the right
/// message when (rarely) more than one request is in flight.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum ChatEvent {
    Token { stream_id: String, delta: String },
    Done { stream_id: String },
    Error { stream_id: String, message: String },
}

// ── Agent ───────────────────────────────────────────────────────────────────

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PlanStatus {
    Pending,
    Running,
    Done,
    Failed,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PlanStep {
    pub id: String,
    pub description: String,
    pub status: PlanStatus,
}

/// A structured tool invocation proposed by the model (L1, validated against the
/// tool's JSON schema before execution).
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ToolCall {
    pub call_id: String,
    pub tool: String,
    pub args: serde_json::Value,
    /// True for side-effecting tools that require human approval (L4).
    pub sensitive: bool,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ToolResult {
    pub call_id: String,
    pub ok: bool,
    pub output: String,
}

/// Emitted when a sensitive tool wants to run; the backend awaits an
/// `agent_approve` command carrying the matching `call_id`.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ApprovalRequest {
    pub call_id: String,
    pub tool: String,
    pub args: serde_json::Value,
    pub reason: String,
}

/// Arguments for `agent_start`.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct AgentRequest {
    pub host: String,
    pub port: u16,
    pub model: String,
    pub task: String,
}

/// Arguments for `agent_approve`.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ApprovalDecision {
    pub agent_id: String,
    pub call_id: String,
    pub approved: bool,
}

/// Streamed on [`AGENT_EVENT`]. `agent_id` identifies the originating agent or
/// sub-agent so the UI can render a sub-agent tree.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum AgentEvent {
    Started {
        agent_id: String,
        parent: Option<String>,
        role: String,
        task: String,
    },
    Plan {
        agent_id: String,
        steps: Vec<PlanStep>,
    },
    PlanUpdate {
        agent_id: String,
        step_id: String,
        status: PlanStatus,
    },
    Step {
        agent_id: String,
        index: u32,
        thought: String,
    },
    Token {
        agent_id: String,
        delta: String,
    },
    ToolCall {
        agent_id: String,
        call: ToolCall,
    },
    ToolResult {
        agent_id: String,
        result: ToolResult,
    },
    ApprovalRequest {
        agent_id: String,
        request: ApprovalRequest,
    },
    SubAgentSpawned {
        agent_id: String,
        child_id: String,
        role: String,
        task: String,
    },
    Log {
        agent_id: String,
        message: String,
    },
    Done {
        agent_id: String,
        final_text: String,
    },
    Error {
        agent_id: String,
        message: String,
    },
}

// ── Shared structs for remaining tabs ────────────────────────────────────────

// ── Model Library ──
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ScannedModel {
    pub path: String,
    pub filename: String,
    pub size_bytes: u64,
    pub clean_name: String,
    pub use_case: String,
    pub hf_link: String,
    pub github_link: String,
    pub size_info: String,
    pub status: String, // "pending_enrichment", "enriching", "enriched", "failed"
    #[serde(default)]
    pub family: String,
    #[serde(default)]
    pub version: String,
    #[serde(default)]
    pub tags: Vec<String>,
}

// ── Planner / Kanban ──
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum TaskStatus {
    Backlog,
    Todo,
    InProgress,
    Done,
}

impl TaskStatus {
    pub fn as_str(&self) -> &str {
        match self {
            Self::Backlog => "Backlog",
            Self::Todo => "Todo",
            Self::InProgress => "In Progress",
            Self::Done => "Done",
        }
    }
    pub fn from_str(s: &str) -> Self {
        match s {
            "Backlog" => Self::Backlog,
            "In Progress" => Self::InProgress,
            "Done" => Self::Done,
            _ => Self::Todo,
        }
    }
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct KanbanTask {
    pub id: String,
    pub title: String,
    pub summary: String,
    pub assigned_agent: String,
    pub status: TaskStatus,
    pub priority: String, // "Low", "Medium", "High"
    pub created_at: String,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct PlannerState {
    pub tasks: Vec<KanbanTask>,
}

// ── Agent Monitor ──
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct AgentStatus {
    pub name: String,
    pub current_action: String,
    pub status: String, // "Idle", "Active", "Offline"
    pub last_active: String,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct AgentActivityEvent {
    pub timestamp: String,
    pub agent_name: String,
    pub message: String,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct MonitorState {
    pub agents: Vec<AgentStatus>,
    pub events: Vec<AgentActivityEvent>,
}

// ── Calendar ──
#[derive(Serialize, Deserialize, Clone, Copy, Debug, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum EventStatus {
    Pending,
    Firing,
    Completed,
    Failed,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct CalendarEvent {
    pub id: String,
    pub title: String,
    pub time: String,   // Format: "YYYY-MM-DDTHH:MM:SS"
    pub prompt: String, // Prompt to fire when the event time is reached
    pub note: Option<String>,
    pub color: Option<String>,
    pub status: EventStatus,
    pub result: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct CalendarState {
    pub events: Vec<CalendarEvent>,
}

// ── Instance Monitor ──
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct LlamaInstance {
    pub hostname: String,
    pub port: u16,
    pub status: String, // "online" | "offline" | "checking"
    pub models: Vec<String>,
}

// ── Downloader / Benchmark / Deep Research ──
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct DownloadStatus {
    pub is_downloading: bool,
    pub log: String,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct BenchmarkResult {
    pub pp: String,
    pub tg: String,
    pub pp_speed: String,
    pub tg_speed: String,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct BenchmarkOutput {
    pub raw: String,
    pub results: Vec<BenchmarkResult>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ResearchStatus {
    pub is_running: bool,
    pub log: String,
    pub report: String,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ResearchReportInfo {
    pub filename: String,
    pub path: String,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct OptimizationSuggestion {
    pub key: String,
    pub label: String,
    pub current: String,
    pub recommended: String,
    pub reason: String,
    pub selected: bool,
}
#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq)]
pub struct Memory {
    pub id: String,
    pub title: String,
    pub created_at: String,
    pub tags: Vec<String>,
    pub links: Vec<String>,
    pub scope: String, // "global" or "project"
    pub content: String,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq)]
pub struct SkillOrAgentFile {
    pub name: String,
    pub path: String,
    pub is_skill: bool,
    pub content: String,
    pub source: String,
}
